File: src/core/text_analysis.py
"""
Text analysis utilities — readability metrics, sentence splitting, syllable counting.

All functions are pure, stateless, and operate on plain strings.
Used by usage_stats (backend) and mirrored in frontend/src/lib/textAnalysis.ts.
"""

from __future__ import annotations

import math
import re

# ---------------------------------------------------------------------------
# Syllable estimation
# ---------------------------------------------------------------------------

# Vowel groups in a word approximate syllable count.  Not perfect,
# but good enough for Flesch-Kincaid without dragging in NLTK.
_VOWELS = re.compile(r"[aeiouy]+", re.IGNORECASE)
_SILENT_E = re.compile(r"[^l]e$", re.IGNORECASE)
_TRAILING_ES = re.compile(r"(?:es|ed)$", re.IGNORECASE)


def estimate_syllables(word: str) -> int:
    """Estimate syllable count for an English word.

    Uses a simple vowel-group heuristic with common corrections.
    Every word has at least one syllable.
    """
    w = word.strip().lower()
    if not w:
        return 0
    # Remove trailing punctuation
    w = re.sub(r"[^a-z]", "", w)
    if not w:
        return 0

    count = len(_VOWELS.findall(w))

    # Silent-e at end (e.g. "make" = 1 syllable, not 2)
    if _SILENT_E.search(w):
        count -= 1

    # Trailing -es/-ed that don't add a syllable (e.g. "baked", "makes")
    if _TRAILING_ES.search(w) and len(w) > 3:
        count -= 1

    return max(count, 1)


# ---------------------------------------------------------------------------
# Sentence splitting
# ---------------------------------------------------------------------------

_SENTENCE_BOUNDARY = re.compile(r"(?<=[.!?])\s+")


def split_sentences(text: str) -> list[str]:
    """Split text into sentences on .!? boundaries.

    If no sentence-ending punctuation is found, treat the entire text
    as one sentence.  Whisper output almost always has punctuation.
    """
    if not text or not text.strip():
        return []
    sentences = [s.strip() for s in _SENTENCE_BOUNDARY.split(text.strip()) if s.strip()]
    return sentences if sentences else [text.strip()]


# ---------------------------------------------------------------------------
# Flesch-Kincaid Grade Level
# ---------------------------------------------------------------------------


def flesch_kincaid_grade(text: str) -> float:
    """Compute the Flesch-Kincaid Grade Level for a block of text.

    Returns 0.0 for empty or degenerate input.
    """
    sentences = split_sentences(text)
    if not sentences:
        return 0.0

    words = text.split()
    if not words:
        return 0.0

    total_syllables = sum(estimate_syllables(w) for w in words)
    word_count = len(words)
    sentence_count = len(sentences)

    grade = 0.39 * (word_count / sentence_count) + 11.8 * (total_syllables / word_count) - 15.59
    return round(min(max(grade, 0.0), 20.0), 1)


# ---------------------------------------------------------------------------
# Aggregate text metrics
# ---------------------------------------------------------------------------


def compute_text_metrics(text: str) -> dict:
    """Compute a full metrics dict for a block of text.

    Returns:
        word_count: int
        sentence_count: int
        avg_sentence_length: float  (words per sentence)
        avg_word_length: float  (characters per word)
        long_word_ratio: float  (words > 6 chars / total)
        fk_grade: float  (Flesch-Kincaid grade level)
        syllables_per_word: float
    """
    if not text or not text.strip():
        return {
            "word_count": 0,
            "sentence_count": 0,
            "avg_sentence_length": 0.0,
            "avg_word_length": 0.0,
            "long_word_ratio": 0.0,
            "fk_grade": 0.0,
            "syllables_per_word": 0.0,
        }

    words = text.split()
    word_count = len(words)
    sentences = split_sentences(text)
    sentence_count = len(sentences)

    # Clean words for character-level metrics
    cleaned = [re.sub(r"[^a-zA-Z]", "", w) for w in words]
    cleaned = [c for c in cleaned if c]

    total_chars = sum(len(c) for c in cleaned)
    long_words = sum(1 for c in cleaned if len(c) > 6)
    total_syllables = sum(estimate_syllables(w) for w in words)

    avg_word_length = total_chars / len(cleaned) if cleaned else 0.0
    long_word_ratio = long_words / len(cleaned) if cleaned else 0.0

    return {
        "word_count": word_count,
        "sentence_count": sentence_count,
        "avg_sentence_length": round(word_count / sentence_count, 1) if sentence_count else 0.0,
        "avg_word_length": round(avg_word_length, 1),
        "long_word_ratio": round(long_word_ratio, 3),
        "fk_grade": flesch_kincaid_grade(text),
        "syllables_per_word": round(total_syllables / word_count, 2) if word_count else 0.0,
    }


# ---------------------------------------------------------------------------
# Standard deviation helper
# ---------------------------------------------------------------------------


def std_dev(values: list[float]) -> float:
    """Population standard deviation.  Returns 0.0 for empty or single-element lists."""
    if len(values) < 2:
        return 0.0
    mean = sum(values) / len(values)
    variance = sum((x - mean) ** 2 for x in values) / len(values)
    return round(math.sqrt(variance), 2)

File: frontend/src/lib/textAnalysis.ts
/**
 * Text analysis utilities — readability metrics, syllable counting, etc.
 *
 * TypeScript mirror of src/core/text_analysis.py so the frontend can compute
 * per-transcript metrics client-side from raw_text and normalized_text.
 */

// ---------------------------------------------------------------------------
// Syllable estimation
// ---------------------------------------------------------------------------

const VOWEL_GROUPS = /[aeiouy]+/gi;
const SILENT_E = /[^l]e$/i;
const TRAILING_ES_ED = /(?:es|ed)$/i;

/** Estimate syllable count for an English word (heuristic). */
export function estimateSyllables(word: string): number {
    let w = word.trim().toLowerCase().replace(/[^a-z]/g, "");
    if (!w) return 0;

    const matches = w.match(VOWEL_GROUPS);
    let count = matches ? matches.length : 0;

    if (SILENT_E.test(w)) count--;
    if (TRAILING_ES_ED.test(w) && w.length > 3) count--;

    return Math.max(count, 1);
}

// ---------------------------------------------------------------------------
// Sentence splitting
// ---------------------------------------------------------------------------

const SENTENCE_BOUNDARY = /(?<=[.!?])\s+/;

/** Split text into sentences. Falls back to one sentence if no punctuation. */
export function splitSentences(text: string): string[] {
    const trimmed = text?.trim();
    if (!trimmed) return [];
    const sentences = trimmed.split(SENTENCE_BOUNDARY).filter((s) => s.trim());
    return sentences.length ? sentences : [trimmed];
}

// ---------------------------------------------------------------------------
// Flesch-Kincaid Grade Level
// ---------------------------------------------------------------------------

/** Compute Flesch-Kincaid Grade Level. Returns 0 for empty/degenerate input. */
export function fleschKincaidGrade(text: string): number {
    const sentences = splitSentences(text);
    if (!sentences.length) return 0;

    const words = text.split(/\s+/).filter(Boolean);
    if (!words.length) return 0;

    const totalSyllables = words.reduce((s, w) => s + estimateSyllables(w), 0);
    const grade = 0.39 * (words.length / sentences.length) + 11.8 * (totalSyllables / words.length) - 15.59;

    return Math.round(Math.min(Math.max(grade, 0), 20) * 10) / 10;
}

// ---------------------------------------------------------------------------
// Filler word detection
// ---------------------------------------------------------------------------

const FILLER_SINGLE = new Set([
    "um",
    "uh",
    "uhm",
    "umm",
    "er",
    "err",
    "like",
    "basically",
    "literally",
    "actually",
    "so",
    "well",
    "right",
    "okay",
]);
const FILLER_MULTI = ["you know", "i mean", "kind of", "sort of"];

/** Count filler words and phrases in a string. */
export function countFillers(text: string): number {
    if (!text) return 0;
    const lower = text.toLowerCase();
    let count = 0;

    // Multi-word fillers
    for (const phrase of FILLER_MULTI) {
        let idx = 0;
        while ((idx = lower.indexOf(phrase, idx)) !== -1) {
            count++;
            idx += phrase.length;
        }
    }

    // Single-word fillers
    for (const w of lower.split(/\s+/)) {
        const c = w.replace(/^[.,!?;:'"()\[\]{}]+|[.,!?;:'"()\[\]{}]+$/g, "");
        if (FILLER_SINGLE.has(c)) count++;
    }

    return count;
}

// ---------------------------------------------------------------------------
// Composite text metrics
// ---------------------------------------------------------------------------

export interface TextMetrics {
    wordCount: number;
    sentenceCount: number;
    avgSentenceLength: number;
    avgWordLength: number;
    longWordRatio: number;
    fkGrade: number;
    fillerCount: number;
    fillerDensity: number;
}

/** Compute all text metrics for a block of text. */
export function computeTextMetrics(text: string): TextMetrics {
    if (!text?.trim()) {
        return {
            wordCount: 0,
            sentenceCount: 0,
            avgSentenceLength: 0,
            avgWordLength: 0,
            longWordRatio: 0,
            fkGrade: 0,
            fillerCount: 0,
            fillerDensity: 0,
        };
    }

    const words = text.split(/\s+/).filter(Boolean);
    const wc = words.length;
    const sentences = splitSentences(text);
    const sc = sentences.length;

    const cleaned = words.map((w) => w.replace(/[^a-zA-Z]/g, "")).filter(Boolean);
    const totalChars = cleaned.reduce((s, c) => s + c.length, 0);
    const longWords = cleaned.filter((c) => c.length > 6).length;
    const fillers = countFillers(text);

    return {
        wordCount: wc,
        sentenceCount: sc,
        avgSentenceLength: sc ? Math.round((wc / sc) * 10) / 10 : 0,
        avgWordLength: cleaned.length ? Math.round((totalChars / cleaned.length) * 10) / 10 : 0,
        longWordRatio: cleaned.length ? Math.round((longWords / cleaned.length) * 1000) / 1000 : 0,
        fkGrade: fleschKincaidGrade(text),
        fillerCount: fillers,
        fillerDensity: wc ? Math.round((fillers / wc) * 10000) / 10000 : 0,
    };
}

// ---------------------------------------------------------------------------
// Statistics helpers
// ---------------------------------------------------------------------------

/** Population standard deviation. Returns 0 for < 2 values. */
export function stdDev(values: number[]): number {
    if (values.length < 2) return 0;
    const mean = values.reduce((s, v) => s + v, 0) / values.length;
    const variance = values.reduce((s, v) => s + (v - mean) ** 2, 0) / values.length;
    return Math.round(Math.sqrt(variance) * 100) / 100;
}

/** Build a normal distribution curve from raw values for visualization. */
export function buildDistributionCurve(
    values: number[],
    bucketCount = 20,
): { x: number; y: number }[] {
    if (values.length < 2) return [];

    const mean = values.reduce((s, v) => s + v, 0) / values.length;
    const sd = stdDev(values);
    if (sd === 0) return [{ x: mean, y: 1 }];

    // Use μ ± 3σ for the range (covers 99.7% of the distribution)
    // clamped so x never goes below 0 for count-like data.
    const lo = Math.max(0, mean - 3 * sd);
    const hi = mean + 3 * sd;
    const range = hi - lo || 1;
    const step = range / bucketCount;

    const points: { x: number; y: number }[] = [];
    for (let i = 0; i <= bucketCount; i++) {
        const x = lo + i * step;
        // Gaussian PDF
        const exp = -0.5 * ((x - mean) / sd) ** 2;
        const y = (1 / (sd * Math.sqrt(2 * Math.PI))) * Math.E ** exp;
        points.push({ x: Math.round(x * 10) / 10, y });
    }

    return points;
}
